pysap/plugins/mri/opt.py

```diff
@@ -0,0 +1,389 @@
+                              weights=weights,
+                              thresh_type='soft')
+
+    # Define cost operator, optional
+    if cost == 'auto':
+        cost = costObj([gradient_op, prox_op], tolerance=tol, verbose=verbose)
+
+    # Use or not Nesterov's acceleration in ISTA algorithm
+    if not acceleration:
+        def lambda_update(lambda_param_old): return 1.0
+    else:
+        lambda_update = 'fista'
+
+    alpha_final = opt_fb(gradient_op=gradient_op,
+                         prox_op=prox_op,
+                         x_init=alpha_init,
+                         beta_param=gradient_op.inv_spec_rad,
+                         lambda_update=lambda_update,
+                         max_iter=max_iter,
+                         cost=cost,
+                         verbose=verbose,
+                         **kwargs)
+
+    x_final = linear_op.adj_op(alpha_final)
+    return x_final
+
+
+def rec_condat_vu_2d_p(data, wavelet_name, samples, mu, s_maps, nb_scale=4,
+                       max_iter=150, tol=1e-4, cartesian_sampling=True,
+                       uniform_data_shape=None, cost=None, **kwargs):
+    """
+    Solve the analysis problem :
+    min 0.5 ||Fx - y||^2 + lambda * ||Psi x||_1
+    using Condat-Vu algorithm
+
+    :param data: np.ndarray
+    Data in the Fourier domain
+    :param wavelet_name: str
+    Wavelet transform name, should be in pysap.AVAILABLE_TRANSFORMS
+    :param samples: np.ndarray
+    Position of k_space samples
+    :param mu: float
+    regularization parameter
+    :param s_maps: np.ndarray
+    sensitivity maps
+    :param nb_scale: int, default 4
+    number of decomposition scales
+    :param max_iter: int, default 150
+    Number of iterations
+    :param tol: float, default 1e-4
+    tolerance of convergence, used if cost is auto
+    :param cartesian_sampling: bool
+    Specify if sampling is cartesian or not
+    :param uniform_data_shape: tuple
+    image shape, needed if sampling is not cartesian
+    :param cost: cost method
+    either None, 'auto' or a CostObj instance
+    :param kwargs: extra parameters
+
+    :return: x_final: np.ndarray
+    The reconstructed image
+    :return: alpha_final: np.ndarray
+    The reconstructed dual variable
+    """
+    # TODO add std estimation and reweightings
+    verbose = kwargs.get('verbose', 1)
+    linear_op = Wavelet2(wavelet_name=wavelet_name,
+                         nb_scale=nb_scale)
+    if cartesian_sampling:
+        data_shape = data.shape[0:2]
+        fourier_op = FFT2(samples=samples, shape=data_shape)
+    elif uniform_data_shape is None:
+        raise AttributeError('If using the NUFFT, the uniform_data_shape '
+                             'argument should be provided')
+    else:
+        data_shape = uniform_data_shape
+        fourier_op = NFFT2(samples=samples, shape=data_shape)
+
+    gradient_op = Grad2D_pMRI(data=data, fourier_op=fourier_op, S=s_maps)
+    gradient_op.spec_rad = 1.1
```

Comment:
I don't understand this linen either the next one

Reply:
Fixed, my bad I forgot to remove it